R/FilteredData.R

```diff
@@ -1223,6 +1262,162 @@ FilteredData <- R6::R6Class( # nolint
+    # Creates UI elements for each captured snapshot.
+    # @details
+    # Snapshot is a `teal_slices` converted to a list of list of lists.
+    # Attributes other than class remain unchanged.
+    # Snapshots are added on user request and stored in a `reactiveVal`.
+    # Initial state application is the first snapshot.
+    # Snapshots are displayed in a table that shows the name, an select button and a save button.
+    # Selecting rebuilds `teal_slices` from snapshot, clears filter states and sets new state.
+    # Saving rebuilds `teal_slices` from snapshot and saves it to json file.
+    # Initial state is not explicitly enumerated in table, has separate reset button.
+    srv_snapshot_manager = function(id) {
+      moduleServer(id, function(input, output, session) {
+        ns <- session$ns
+
+        # Store global filter states.
+        snapshot_history <- reactiveVal({
+          list(
+            "Initial application state" = disassemble_slices(isolate(self$get_filter_state()))
+          )
+        })
+
+        # Snapshot current application state - name snaphsot.
+        observeEvent(input$snapshot_add, {
+          showModal(
+            modalDialog(
+              textInput(
+                session$ns("snapshot_name"),
+                "Name the snapshot",
+                width = "100%",
+                placeholder = "Meaningful, unique name"
+              ),
+              footer = tagList(
+                actionButton(session$ns("snapshot_name_accept"), "Accept", icon = icon("thumbs-up")),
+                modalButton(label = "Cancel", icon = icon("thumbs-down"))
+              ),
+              size = "s"
+            )
+          )
+        })
+        # Snapshot current application state - store snaphsot.
+        observeEvent(input$snapshot_name_accept, {
+          snapshot_name <- trimws(input$snapshot_name)
+          if (identical(snapshot_name, "")) {
+            showNotification(
+              "Please name the snapshot.",
+              type = "message"
+            )
+            updateTextInput(inputId = "snapshot_name", value = "", placeholder = "Meaningful, unique name")
+          } else if (is.element(make.names(snapshot_name), make.names(names(snapshot_history())))) {
+            showNotification(
+              "This name is in conflict with other snapshot names. Please choose a different one.",
+              type = "message"
+            )
+            updateTextInput(inputId = "snapshot_name", value = , placeholder = "Meaningful, unique name")
+          } else {
+            snapshot <- disassemble_slices(self$get_filter_state())
+            snapshot_update <- c(snapshot_history(), list(snapshot))
+            names(snapshot_update)[length(snapshot_update)] <- snapshot_name
+            snapshot_history(snapshot_update)
+            removeModal()
+            shinyjs::click(id = "teal-main_ui-filter_panel-active-show_snapshot_manager", asis = TRUE)
```

Comment:
I'd prefer modal to be closed with notification instead or reopening with js shot based on the id which package has no control of 
```suggestion
```

Reply:
The request was that first modal (snapshot manager) remain open to facilitate immediate further actions but it is closed as the second one (snapshot name) is opened. I agree that this method is not ideal to say the least.

Reply:
Line doesn't guarantee that modal will be open as we might change `teal` namespaces and we would never realise that it doesn't work anymore. This line has a dependency on teal (reverse) and is impossible to test. The request is understandable but we can't guarantee this to work in a long term. 
@lcd2yyz what do you think if alternatively we close the window and show a notification that "Snapshot has been added"

Reply:
Thanks for explaining. While I do want to have the re-open action, future reliability is also important, so your proposal sounds reasonable compromise to me.